### ISSUE TITLE
Vector support for generic call

### ISSUE THREAD

**Issue by stelmo** [EnzymeAD/Enzyme.jl#304]
Hi, I am trying to find the Jacobian of a vector function, but no matter what I try, `jacobian` crashes Julia. A minimal working example is shown below:
```julia
using Enzyme
using SparseArrays, LinearAlgebra

N = 10
_S = sprand(N, N, 0.1) 
_Q = sprand(N, N, 0.1)

F(z) = [ # some complicated sparse amenable function
    _S * z 
    _Q * z.^2
]
z = rand(N)

jac = Enzyme.jacobian(Forward, F, z) # crash

# Dense version also fails

N = 10
_S = rand(N, N) 
_Q = rand(N, N)

F(z) = [
    _S * z 
    _Q * z.^2
]
z = rand(N)

jac = Enzyme.jacobian(Forward, F, z) # crash
```
I am using this `[7da242da] Enzyme v0.10.0-dev https://github.com/EnzymeAD/Enzyme.jl.git#main` version of Enzyme, which should be the latest. Is there a workaround or should I just be patient? :)

**Comment by vchuravy** [comment]
Can you post the full log as a Gist?

**Comment by wsmoses** [comment]
The issue here is that we haven’t added vector mode to generic calls yet.
Can you specify a chunk size of 1

**Comment by vchuravy** [comment]
`jac = Enzyme.jacobian(Forward, F, z, Val(1))`

**Comment by vchuravy** [comment]
Note that what You might want to make your code type-stable.

```
julia> const S = sprand(N, N, 0.1)
10×10 SparseMatrixCSC{Float64, Int64} with 10 stored entries:
  ⋅   0.864191   ⋅         ⋅         ⋅        0.725971   ⋅         ⋅    ⋅         ⋅ 
  ⋅    ⋅         ⋅        0.454994   ⋅         ⋅         ⋅         ⋅   0.404086   ⋅ 
  ⋅    ⋅         ⋅         ⋅         ⋅        0.937271  0.321421   ⋅    ⋅         ⋅ 
  ⋅    ⋅         ⋅         ⋅        0.325655   ⋅         ⋅         ⋅    ⋅         ⋅ 
  ⋅   0.354443   ⋅         ⋅         ⋅         ⋅         ⋅         ⋅    ⋅         ⋅ 
  ⋅    ⋅        0.398672   ⋅         ⋅         ⋅         ⋅         ⋅    ⋅         ⋅ 
  ⋅    ⋅         ⋅         ⋅         ⋅         ⋅        0.14057    ⋅    ⋅         ⋅ 
  ⋅    ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅    ⋅         ⋅ 
  ⋅    ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅    ⋅         ⋅ 
  ⋅    ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅    ⋅         ⋅ 

julia> const Q = _Q
10×10 SparseMatrixCSC{Float64, Int64} with 6 stored entries:
  ⋅    ⋅    ⋅    ⋅        0.450911   ⋅         ⋅    ⋅   0.0822345   ⋅ 
  ⋅    ⋅    ⋅    ⋅         ⋅         ⋅         ⋅    ⋅    ⋅          ⋅ 
  ⋅    ⋅    ⋅    ⋅         ⋅         ⋅         ⋅    ⋅    ⋅          ⋅ 
  ⋅    ⋅    ⋅    ⋅         ⋅         ⋅         ⋅    ⋅    ⋅          ⋅ 
  ⋅    ⋅    ⋅    ⋅         ⋅         ⋅         ⋅    ⋅    ⋅          ⋅ 
  ⋅    ⋅    ⋅    ⋅         ⋅         ⋅         ⋅    ⋅    ⋅          ⋅ 
  ⋅    ⋅    ⋅    ⋅         ⋅         ⋅         ⋅    ⋅    ⋅          ⋅ 
  ⋅    ⋅    ⋅    ⋅         ⋅         ⋅         ⋅    ⋅   0.712695    ⋅ 
  ⋅    ⋅    ⋅    ⋅         ⋅        0.670415   ⋅    ⋅    ⋅          ⋅ 
  ⋅    ⋅    ⋅   0.834334   ⋅        0.719653   ⋅    ⋅    ⋅          ⋅ 

julia> F(z) = [ # some complicated sparse amenable function
           S * z 
           Q * z.^2
       ]
F (generic function with 2 methods)

julia> jac = Enzyme.jacobian(Forward, F, z)
(([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.8641913606126247, 0.0, 0.0, 0.0, 0.3544431260542895, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.39867161256143524, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, 0.4549936171986305, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.14561249567530618], [0.0, 0.0, 0.0, 0.3256549706212486, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.8075284885092465, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.7259714923753596, 0.0, 0.9372709878409331, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0423793826488412, 1.1189362287523177], [0.0, 0.0, 0.32142094179075853, 0.0, 0.0, 0.0, 0.14057035792574357, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, 0.4040861036618695, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.05656868148254788, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.4902595563499479, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]),)
```

**Comment by vchuravy** [comment]
In particular the type-stable variant:

```
julia> @btime  Enzyme.jacobian(Forward, F, z)
  2.236 μs (59 allocations: 8.86 KiB)
```

Versus your variant:

```
julia> @btime  Enzyme.jacobian(Forward, F, z, Val(1))
  115.640 μs (855 allocations: 56.03 KiB)
```

**Comment by wsmoses** [comment]
For the time for curiousity can you also try the type stable, but vector size 1 as well

**Comment by vchuravy** [comment]
```
julia> @btime  Enzyme.jacobian(Forward, F, z, Val(1))
  3.336 μs (105 allocations: 14.78 KiB)

julia> @btime  Enzyme.jacobian(Forward, F, z)
  2.212 μs (59 allocations: 8.86 KiB)
```

**Comment by stelmo** [comment]
This worked! Thanks :) Although I ran into another issue, but I am not sure if they are related
```julia 
using Enzyme
using SparseArrays, LinearAlgebra
using Random 

N = 10
n = 2
rs = shuffle(1:N)[1:round(Int, N/n)]
cs = shuffle(1:N)[1:round(Int, N/n)]
_Q = sprand(N, N, 0.1)
_S(x) = sparse(rs, cs, x, N, N)

z = rand(N)
x = rand(round(Int, N/n))
_S(x)

F(x, z) = [ # some complicated sparse amenable function
    _S(x) * z 
    _Q * z.^2
]

F(x, z)

jac = Enzyme.jacobian(Forward, z -> F(x, z), z, Val(1))
```
causes this error:
```julia
julia> jac = Enzyme.jacobian(Forward, z -> F(x, z), z, Val(1))
ERROR: AssertionError: length(args) == length((collect(parameters(entry_f)))[1 + sret:end])
Stacktrace:
 [1] lower_convention(functy::Type, mod::LLVM.Module, entry_f::LLVM.Function, actualRetType::Type)
   @ Enzyme.Compiler C:\Users\stelmo\.julia\packages\Enzyme\Ctome\src\compiler.jl:3553
 [2] codegen(output::Symbol, job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(sparse), Tuple{Vector{Int64}, Vector{Int64}, Vector{Float64}, Int64, Int64}}}; libraries::Bool, deferred_codegen::Bool, optimize::Bool, ctx::LLVM.Context, strip::Bool, validate::Bool, only_entry::Bool, parent_job::Nothing)
   @ Enzyme.Compiler C:\Users\stelmo\.julia\packages\Enzyme\Ctome\src\compiler.jl:3958
 [3] _thunk(job::GPUCompiler.CompilerJob{Enzyme.Compiler.EnzymeTarget, Enzyme.Compiler.EnzymeCompilerParams, GPUCompiler.FunctionSpec{typeof(sparse), Tuple{Vector{Int64}, Vector{Int64}, Vector{Float64}, Int64, Int64}}})
   @ Enzyme.Compiler C:\Users\stelmo\.julia\packages\Enzyme\Ctome\src\compiler.jl:4397
 [4] cached_compilation(job::GPUCompiler.CompilerJob, key::UInt64, specid::UInt64)
   @ Enzyme.Compiler C:\Users\stelmo\.julia\packages\Enzyme\Ctome\src\compiler.jl:4435
 [5] #s501#108
   @ C:\Users\stelmo\.julia\packages\Enzyme\Ctome\src\compiler.jl:4495 [inlined]
 [6] var"#s501#108"(F::Any, DF::Any, A::Any, TT::Any, Mode::Any, ModifiedBetween::Any, width::Any, specid::Any, ::Any, f::Any, df::Any, #unused#::Type, tt::Any, #unused#::Type, #unused#::Type, #unused#::Type, #unused#::Any)
   @ Enzyme.Compiler .\none:0
 [7] (::Core.GeneratedFunctionStub)(::Any, ::Vararg{Any})
   @ Core .\boot.jl:580
 [8] thunk
   @ C:\Users\stelmo\.julia\packages\Enzyme\Ctome\src\compiler.jl:4523 [inlined]
 [9] thunk(f::typeof(sparse), df::Nothing, ::Type{Duplicated}, tt::Type{Tuple{Const{Vector{Int64}}, Const{Vector{Int64}}, Const{Vector{Float64}}, Const{Int64}, Const{Int64}}}, ::Val{Enzyme.API.DEM_ForwardMode}, ::Val{0x0000000000000001})
   @ Enzyme.Compiler C:\Users\stelmo\.julia\packages\Enzyme\Ctome\src\compiler.jl:4516
```

**Comment by wsmoses** [comment]
The first code now hits a:
```
julia: /buildworker/worker/package_linux64/build/src/llvm-late-gc-lowering.cpp:870: std::vector<int, std::allocator<int> > LateLowerGCFrame::NumberAllBase(State&, llvm::Value*): Assertion `Tracked.size() == BaseNumbers.size()' failed.

signal (6): Aborted
in expression starting at REPL[8]:1
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7f2c0a4d53f9)
__assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
NumberAllBase at /buildworker/worker/package_linux64/build/src/llvm-late-gc-lowering.cpp:870
NumberBase at /buildworker/worker/package_linux64/build/src/llvm-late-gc-lowering.cpp:787
Number at /buildworker/worker/package_linux64/build/src/llvm-late-gc-lowering.cpp:804
LocalScan at /buildworker/worker/package_linux64/build/src/llvm-late-gc-lowering.cpp:1590
runOnFunction at /buildworker/worker/package_linux64/build/src/llvm-late-gc-lowering.cpp:2675
_ZN4llvm13FPPassManager13runOnFunctionERNS_8FunctionE at /mnt/Data/git/Enzyme.jl/julia-1.7.2/bin/../lib/julia/libLLVM-12jl.so (unknown line)
```

Your latest code hits:
```
julia> Enzyme.API.looseTypeAnalysis!(true)

julia> jac = Enzyme.jacobian(Forward, z -> F(x, z), z, Val(1))
┌ Warning: Returned rooting not fully handled, segfault likely
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/jVY4I/src/utils.jl:35
warning: binary operator is integer and constant:   %59 = add i64 %58, 1, !dbg !95
warning: binary operator is integer and constant:   %155 = add i64 %154, 1, !dbg !201
ERROR: UndefRefError: access to undefined reference
Stacktrace:
 [1] getindex(t::Tuple, i::Int64)
   @ Base ./tuple.jl:29
   ```

**Comment by wsmoses** [comment]
First code remains a GC issue of vector mode bug (the same as above).

The second code is now a segfault:
```
julia> jac = Enzyme.jacobian(Forward, z -> F(x, z), z, Val(1))


┌ Warning: Returned rooting not fully handled, segfault likely
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/07qaN/src/utils.jl:35

signal (11): Segmentation fault
in expression starting at REPL[15]:1
unknown function (ip: 0x7ff5d9833410)
unknown function (ip: 0x7ff5d9833730)
Allocations: 71962723 (Pool: 71893824; Big: 68899); GC: 76
Segmentation fault (core dumped)
```

**Comment by wsmoses** [comment]
Original code now fixed, upon use of 1.8.3 with the vector bug fixed in Julia proper, the second code now also appears to work, so closing.

**Comment by stelmo** [comment]
Thanks to both of you for fixing these issues. I really appreciate the effort you put into this package, it is great!!
